### PR TITLE
Optimized find pooled resources in ZonalParkingManager

### PR DESCRIPTION
**Pull request by REASY**
This change is [Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/744)

### CODE REVIEW

File: src/main/scala/beam/agentsim/infrastructure/ZonalParkingManager.scala [hunk 1]
```diff
@@ -250,6 +241,62 @@ class ZonalParkingManager(
...
+          }
+      }
+    }
+    def predicate(tazId: Id[TAZ], preferredType: ParkingType, pricingModel: PricingModel,
+                  reservedFor: ReservedParkingType): Boolean = {
+      val key = IndexForFind(tazId = tazId, parkingType = preferredType, pricingModel = pricingModel, reservedFor = reservedFor)
+      find(key).isDefined
+    }
+    // We still have to do two look-ups by key because we don't know exact pricing model
+    allPricingModels.collectFirst { case pricingModel if predicate(tazId, preferredType, pricingModel, reservedFor) =>
```
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/infrastructure/ZonalParkingManager.scala]: You'll save yourself some allocations by moving the predicate into the logic. Then you can build one object, check if it's defined and get it. However you have to use something more like:

```
allPricingModels
  .view
  .map(pricingModel => find(IndexForFind(tazId, preferredType, pricingModel, reservedFor))
  .filter(_.isDefined)
  .head.get
```
And, you could do this for the other predicate also.

File: src/main/scala/beam/agentsim/infrastructure/ZonalParkingManager.scala [hunk 2]
```diff
@@ -45,8 +59,11 @@ class ZonalParkingManager(
     NoCharger,
     ParkingStall.Any
   )
-  val defaultStallValues = StallValues(Int.MaxValue, 0)
-  val defaultRideHailStallValues = StallValues(0, 0)
+  def defaultStallValues: StallValues = StallValues(Int.MaxValue, 0)
+  def defaultRideHailStallValues: StallValues = StallValues(0, 0)
+
+  // TODO How to make sure that this will be in sync? One of the solution is to make runtime check using reflection
+  val allPricingModels: Array[PricingModel] = Array(FlatFee, Block)
```
[the remark this answers is not in the document]
Comment (JustinPihony) [reply src/main/scala/beam/agentsim/infrastructure/ZonalParkingManager.scala]: Also, this seems to be an addition to the current logic. Does this make the logic better, fix something, etc? Or maybe I missed something.

**Comment by JustinPihony** [comment]
I just now noticed the WIP - so feel free to ignore my comments :p

File: src/main/scala/beam/agentsim/infrastructure/ZonalParkingManager.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (REASY) [reply src/main/scala/beam/agentsim/infrastructure/ZonalParkingManager.scala]: it's just for the purpose of optimization. We create two indexes `IndexForFind` and `IndexForFilter` to be able to reduce linear time to amortized constant time for filter and find.

File: src/main/scala/beam/agentsim/infrastructure/ZonalParkingManager.scala (hunk 1 above)
Comment (REASY) [reply src/main/scala/beam/agentsim/infrastructure/ZonalParkingManager.scala]: I thought about this approach, but decided to go with `collectFirst`. Yes, for `collectFirst` I have to do one extra call. But if what you proposed is more readable(and it should be definitely faster), I can make changes

**Comment by colinsheppard** [comment]
@REASY can you initiate a production run (1 iteration is fine) from this branch and post the results link? Thank you!

**Comment by REASY** [comment]
I did run in cluster(master+2workers). Here is output for 2 iterations: https://drive.google.com/open?id=1yLVFVk_5q1O2wTOdx2TsfaysvDxPPfMz
Agent Sim times:
Iteration 0 => 5536 seconds
Iteration 1 => 3416 seconds.

**Comment by REASY** [comment]
Results:
No optimization: https://drive.google.com/open?id=105GaMdED7el7aqkrqK5FPYu6eN9tH_1d
With optimization: https://drive.google.com/open?id=1uOgd90tzWRYSyjbgBaoiBkz2KlmViFT6

**Comment by REASY** [comment]
Call tree for non-optimized version:
image

**Comment by REASY** [comment]
Call tree for optimized one:
image

**Comment by REASY** [comment]
New run with 5 iterations
No optimization: https://drive.google.com/open?id=1dNUe6uqM55NmobqWIQyZv9bWP729os_X
Time:
Iteration 0 => 7169 s
Iteration 1 => 5153 s
Iteration 2 => 5202 s
Iteration 3 => 5350 s
Iteration 4 => 5138 s

**Pull request merged**